# Week-9 Assignment 9.2

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

# 1.Import the data as a data frame and ensure it is loaded correctly.

In [2]:
# Read house price train data csv file 
df_Loan_Train_data=pd.read_csv('Loan_Train.csv')

In [3]:
df_Loan_Train_data.shape

(614, 13)

In [4]:
df_Loan_Train_data.dtypes

Loan_ID               object
Gender                object
Married               object
Dependents            object
Education             object
Self_Employed         object
ApplicantIncome        int64
CoapplicantIncome    float64
LoanAmount           float64
Loan_Amount_Term     float64
Credit_History       float64
Property_Area         object
Loan_Status           object
dtype: object

In [5]:
df_Loan_Train_data.head()

Loan_ID Gender Married Dependents     Education Self_Employed  \
0  LP001002   Male      No          0      Graduate            No   
1  LP001003   Male     Yes          1      Graduate            No   
2  LP001005   Male     Yes          0      Graduate           Yes   
3  LP001006   Male     Yes          0  Not Graduate            No   
4  LP001008   Male      No          0      Graduate            No   

   ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
0             5849                0.0         NaN             360.0   
1             4583             1508.0       128.0             360.0   
2             3000                0.0        66.0             360.0   
3             2583             2358.0       120.0             360.0   
4             6000                0.0       141.0             360.0   

   Credit_History Property_Area Loan_Status  
0             1.0         Urban           Y  
1             1.0         Rural           N  
2             1.0         Urban           Y  
3             1.0         Urban           Y  
4             1.0         Urban           Y

# 2) Prepare the data for modeling by performing the following steps:

1) Drop the column “Load_ID.”

In [6]:
df_Loan_Train_data.drop('Loan_ID',axis=1)

Gender Married Dependents     Education Self_Employed  ApplicantIncome  \
0      Male      No          0      Graduate            No             5849   
1      Male     Yes          1      Graduate            No             4583   
2      Male     Yes          0      Graduate           Yes             3000   
3      Male     Yes          0  Not Graduate            No             2583   
4      Male      No          0      Graduate            No             6000   
..      ...     ...        ...           ...           ...              ...   
609  Female      No          0      Graduate            No             2900   
610    Male     Yes         3+      Graduate            No             4106   
611    Male     Yes          1      Graduate            No             8072   
612    Male     Yes          2      Graduate            No             7583   
613  Female      No          0      Graduate           Yes             4583   

     CoapplicantIncome  LoanAmount  Loan_Amount_Term  Credit_History  \
0                  0.0         NaN             360.0             1.0   
1               1508.0       128.0             360.0             1.0   
2                  0.0        66.0             360.0             1.0   
3               2358.0       120.0             360.0             1.0   
4                  0.0       141.0             360.0             1.0   
..                 ...         ...               ...             ...   
609                0.0        71.0             360.0             1.0   
610                0.0        40.0             180.0             1.0   
611              240.0       253.0             360.0             1.0   
612                0.0       187.0             360.0             1.0   
613                0.0       133.0             360.0             0.0   

    Property_Area Loan_Status  
0           Urban           Y  
1           Rural           N  
2           Urban           Y  
3           Urban           Y  
4           Urban           Y  
..            ...         ...  
609         Rural           Y  
610         Rural           Y  
611         Urban           Y  
612         Urban           Y  
613     Semiurban           N  

[614 rows x 12 columns]

2) Drop any rows with missing data.


In [7]:
df_Loan_Train_data.dropna(inplace=True)
df_Loan_Train_data.isna().sum()

Loan_ID              0
Gender               0
Married              0
Dependents           0
Education            0
Self_Employed        0
ApplicantIncome      0
CoapplicantIncome    0
LoanAmount           0
Loan_Amount_Term     0
Credit_History       0
Property_Area        0
Loan_Status          0
dtype: int64

3) convert the categorical features into dummy variables.

In [8]:
# convert categorical data to_numeric 
df_Loan_Train_data = pd.get_dummies(df_Loan_Train_data, drop_first=True) 
df_Loan_Train_data.head() 


ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
1             4583             1508.0       128.0             360.0   
2             3000                0.0        66.0             360.0   
3             2583             2358.0       120.0             360.0   
4             6000                0.0       141.0             360.0   
5             5417             4196.0       267.0             360.0   

   Credit_History  Loan_ID_LP001005  Loan_ID_LP001006  Loan_ID_LP001008  \
1             1.0                 0                 0                 0   
2             1.0                 1                 0                 0   
3             1.0                 0                 1                 0   
4             1.0                 0                 0                 1   
5             1.0                 0                 0                 0   

   Loan_ID_LP001011  Loan_ID_LP001013  ...  Gender_Male  Married_Yes  \
1                 0                 0  ...            1            1   
2                 0                 0  ...            1            1   
3                 0                 0  ...            1            1   
4                 0                 0  ...            1            0   
5                 1                 0  ...            1            1   

   Dependents_1  Dependents_2  Dependents_3+  Education_Not Graduate  \
1             1             0              0                       0   
2             0             0              0                       0   
3             0             0              0                       1   
4             0             0              0                       0   
5             0             1              0                       0   

   Self_Employed_Yes  Property_Area_Semiurban  Property_Area_Urban  \
1                  0                        0                    0   
2                  1                        0                    1   
3                  0                        0                    1   
4                  0                        0                    1   
5                  1                        0                    1   

   Loan_Status_Y  
1              0  
2              1  
3              1  
4              1  
5              1  

[5 rows x 494 columns]

# 3) Split the data into a training and test set, where the “Loan_Status” column is the target.

In [9]:
from sklearn.model_selection import train_test_split 
X = df_Loan_Train_data.drop('Loan_Status_Y',axis=1)
y = df_Loan_Train_data["Loan_Status_Y"]
X_train, X_test, y_train,y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# 4) Create a pipeline with a min-max scaler and a KNN classifier (see section 15.3 in the Machine Learning with Python Cookbook).

In [10]:
from sklearn.pipeline import Pipeline 
from sklearn.preprocessing import MinMaxScaler 
from sklearn.neighbors import KNeighborsClassifier 
pipe = Pipeline([('scaler', MinMaxScaler()), ('knn',KNeighborsClassifier())]) 

# 5) Fit a default KNN classifier to the data with this pipeline. Report the model accuracy on the test set. Note: Fitting a pipeline model works just like fitting a regular model.

In [11]:
pipe.fit(X_train, y_train) # %% # accuracy of the model
print(pipe.score(X_test, y_test))

0.78125


# 6) Create a search space for your KNN classifier where your “n_neighbors” parameter varies from 1 to 10. (see section 15.3 in the Machine Learning with Python Cookbook).

# 7) Fit a grid search with your pipeline, search space, and 5-fold cross-validation to find the best value for the “n_neighbors” parameter.

In [12]:
# Using the GridSearchCV class to find the best value for 'n_neighbors' parameter.
from sklearn.model_selection import GridSearchCV 
pipe = Pipeline([('scaler', MinMaxScaler()), ('knn',KNeighborsClassifier())]) 
param_grid = {'knn__n_neighbors': range(1, 11)} 
classifier = GridSearchCV(pipe, param_grid, cv=5, n_jobs=-1, verbose=1).fit(X_train, y_train) 

Fitting 5 folds for each of 10 candidates, totalling 50 fits




# 8) Find the accuracy of the grid search best model on the test set. Note: It is possible that this will not be an improvement over the default model, but likely it will be.
# 9)Now, repeat steps 6 and 7 with the same pipeline, but expand your search space to include logistic regression and random forest models with the hyperparameter values in section 12.3 of the Machine Learning with Python Cookbook.

In [13]:
# Best value for n_neighbors classifier.best_params_ 
# %% # Accuracy of the grid search best model on the test set. classifier.best_score_
# %% # repeat steps 6 and 7 with the same pipeline, but expand your search space to include logistic regression and random forest models with the hyperparameter. 
from sklearn.linear_model import LogisticRegression 
from sklearn.ensemble import RandomForestClassifier 
pipe = Pipeline([('scaler', MinMaxScaler()), ('classifier', KNeighborsClassifier())])
search_space = [{ "classifier": [LogisticRegression()], "classifier__penalty":
['l2'], "classifier__C": np.logspace(0, 4, 10) }, { "classifier":
[RandomForestClassifier()], "classifier__n_estimators": [10, 100, 1000],
"classifier__max_features": [1, 2, 3] }, { "classifier": [KNeighborsClassifier()],
"classifier__n_neighbors": range(1, 11), "classifier__weights": ['uniform',
'distance'] }] 
gridsearch = GridSearchCV(pipe, search_space, cv=5, verbose=0, n_jobs=-1) 
gridsearch.fit(X_train, y_train)                                                                                      


GridSearchCV(cv=5,
             estimator=Pipeline(steps=[('scaler', MinMaxScaler()),
                                       ('classifier', KNeighborsClassifier())]),
             n_jobs=-1,
             param_grid=[{'classifier': [LogisticRegression(C=7.742636826811269)],
                          'classifier__C': array([1.00000000e+00, 2.78255940e+00, 7.74263683e+00, 2.15443469e+01,
       5.99484250e+01, 1.66810054e+02, 4.64158883e+02, 1.29154967e+03,
       3.59381366e+03, 1.00000000e+04]),
                          'classifier__penalty': ['l2']},
                         {'classifier': [RandomForestClassifier()],
                          'classifier__max_features': [1, 2, 3],
                          'classifier__n_estimators': [10, 100, 1000]},
                         {'classifier': [KNeighborsClassifier()],
                          'classifier__n_neighbors': range(1, 11),
                          'classifier__weights': ['uniform', 'distance']}])

# 10)What are the best model and hyperparameters found in the grid search? Find the accuracy of this model on the test set.

In [26]:
best_model = gridsearch.fit(X_train, y_train) 
y_pred = best_model.predict(X_test)


In [28]:
#calculate accuracy
from sklearn.metrics import accuracy_score
accuracy_score(y_test,y_pred)

0.8229166666666666

# 11) Summarize your results.

From the above we understand that the KNN model accuracy is only 78%, logistic regression aacuracy is 82% and C value 7.74,which tells us that regression is the best model for the data.